### PR TITLE
Creating unified base class for all topic models, #938

### LINKED ISSUE #938: Create single API/base class for all topic models
As discussed on gitter, it would be nice to have a single API/base class which can be extended by all topic models. This will include common functions such as `print_topic()` which are needed by all models. eg. `show_topic()` is not present in `HdpModel` currently and adding this function would incur code duplication.

### PR DESCRIPTION
**Pull request by markroxor**
Attempt to fix #938. Waiting for moderator review to proceed with further commits.

### CODE REVIEW

File: gensim/models/base.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class BaseModel():
```
Comment (tmylk) [inline gensim/models/base.py]: BaseTopicModel is more appropriate

**Comment by tmylk** [comment]
Please rebase off develop and add HDPModel

**Comment by markroxor** [comment]
`print_topic` is not present in HDPModel. Should I create one like that in LDAModel and LSIModel?

**Comment by markroxor** [comment]
I was not sure about what tests to add, please take a close look on the test files.

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,10 +10,11 @@ Changes
 * Change export_phrases in Phrases model. Fix issue #794 (@AadityaJ,
 [#879](https://github.com/RaRe-Technologies/gensim/pull/879))
     - bigram construction can now support multiple bigrams within one sentence
-* Fixed issue #838, RuntimeWarning: overflow encountered in exp  (@markroxor, [#895](https://github.com/RaRe-Technologies/gensim/pull/895))
```
Comment (tmylk) [inline CHANGELOG.md]: why these changes?
Expect to see just 1 line added about the new fix.

File: gensim/test/test_ldamodel.py [hunk 3]
```diff
@@ -253,6 +255,12 @@ def testShowTopics(self):
                 self.assertTrue(isinstance(k, six.string_types))
                 self.assertTrue(isinstance(v, float))
 
+        #testing print_topic
```
Comment (tmylk) [inline gensim/test/test_ldamodel.py]: create new test test_print_topic in new class TestBaseTopicModel.
TestLdaModel, testLSI and testHDP should inherit from TestBaseTopicModel.

Comment (markroxor) [inline gensim/test/test_ldamodel.py]: Are the tests okay (the assertions of variable type)?

File: CHANGELOG.md (hunk 2 above)
Comment (markroxor) [inline CHANGELOG.md]: I was fixing my previous hyperlinks, but alright. I'll add just one line no issues.

File: gensim/test/test_ldamodel.py (hunk 3 above)
Comment (markroxor) [inline gensim/test/test_ldamodel.py]: Should I also create a unified test_basemodel.py to handle common tests in all the models like testShowTopic and testShowTopics?

File: CHANGELOG.md (hunk 2 above)
Comment (tmylk) [inline CHANGELOG.md]: Fixes to your prev links are ok. You can leave them in.

**Comment by tmylk** [comment]
BTW `HdpTopicFormatter` has the `print_topic` method

File: gensim/models/ldamodel.py [hunk 4]
```diff
@@ -193,7 +195,7 @@ def get_Elogbeta(self):
 # endclass LdaState
 
 
-class LdaModel(interfaces.TransformationABC):
+class LdaModel(interfaces.TransformationABC,basemodel.BaseTopicModel):
```
Comment (dsquareindia) [inline gensim/models/ldamodel.py]: slight code style issue. Should be `interfaces.TransformationABC, basemodel.BaseTopicModel`.

File: gensim/models/lsimodel.py [hunk 5]
```diff
@@ -221,7 +223,7 @@ def merge(self, other, decay=1.0):
 #endclass Projection
 
 
-class LsiModel(interfaces.TransformationABC):
+class LsiModel(interfaces.TransformationABC,basemodel.BaseTopicModel):
```
Comment (dsquareindia) [inline gensim/models/lsimodel.py]: same here. You should probably consider setting up a flake8 checks plugin in your editor. Will help you eliminate those nitpicks.

Comment (markroxor) [inline gensim/models/lsimodel.py]: Thanks for the suggestions. I will keep that in mind.

File: gensim/test/test_ldamodel.py (hunk 3 above)
Comment (tmylk) [inline gensim/test/test_ldamodel.py]: yes

File: gensim/test/test_hdpmodel.py [hunk 6]
```diff
@@ -47,7 +47,6 @@ def testfile():
     return os.path.join(tempfile.gettempdir(), 'gensim_models.tst')
 
 
-
 class TestHdpModel(unittest.TestCase):
```
Comment (tmylk) [inline gensim/test/test_hdpmodel.py]: should it inherit from TestBaseTopicModel too? Feel free to add the missing methods

File: gensim/models/hdpmodel.py [hunk 7]
```diff
@@ -575,9 +570,6 @@ def __init__(self, dictionary=None, topic_data=None, topic_file=None, style=None
 
         self.style = style
 
-    def print_topics(self, num_topics=10, num_words=10):
```
Comment (markroxor) [inline gensim/models/hdpmodel.py]: It is inherited from the BaseTopicModel class just like lsimodel and ldamodel.

Comment (tmylk) [inline gensim/models/hdpmodel.py]: this method is in class HDPFormatter, not HDP

Comment (markroxor) [inline gensim/models/hdpmodel.py]: okay I will restore it.

Comment (tmylk) [inline gensim/models/hdpmodel.py]: It is still deleted.

File: gensim/models/hdpmodel.py [hunk 8]
```diff
@@ -612,10 +604,9 @@ def show_topic_terms(self, topic_data, num_words):
     def format_topic(self, topic_id, topic_terms):
         if self.STYLE_GENSIM == self.style:
             fmt = ' + '.join(['%.3f*%s' % (weight, word) for (word, weight) in topic_terms])
-            fmt = 'topic %i: %s' % (topic_id, fmt)
```
Comment (tmylk) [inline gensim/models/hdpmodel.py]: please revert

**Comment by tmylk** [comment]
Please remove other people's commits from this PR

File: gensim/test/test_basemodel.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        topics = self.model.print_topics()
+
+        for topic_no, topic in topics:
+            self.assertTrue(isinstance(topic_no, int))
+            self.assertTrue(isinstance(topic, str) or isinstance(topic, unicode))
+
+    def testShowTopic(self):
+        # TODO create show_topic in HdpModel and then test
+
+        if self.__class__.__name__ == "TestHdpModel":
```
Comment (tmylk) [inline gensim/test/test_basemodel.py]: please override testShowTopic inside TestHdpModel and skip it there

File: gensim/models/hdpmodel.py [hunk 10]
```diff
@@ -612,10 +604,9 @@ def show_topic_terms(self, topic_data, num_words):
...
         else:
             fmt = '\n'.join(['    %20s    %.8f' % (word, weight) for (word, weight) in topic_terms])
-            fmt = 'topic %i:\n%s' % (topic_id, fmt)
```
Comment (markroxor) [inline gensim/models/hdpmodel.py]: This is for passing the testShowTopics test, all the other models throw data in this format. Should I exclude this test too?

Comment (markroxor) [inline gensim/models/hdpmodel.py]: This is for passing the testShowTopics test, all the other models throw data in this format. Should I exclude this test too?

Comment (tmylk) [inline gensim/models/hdpmodel.py]: it is a different class

Comment (markroxor) [inline gensim/models/hdpmodel.py]: waht about `testShowTopics` test? should I override?

File: gensim/models/hdpmodel.py (hunk 7 above)
Comment (tmylk) [inline gensim/models/hdpmodel.py]: Please revert all changes to HdpTopicFormatter

File: gensim/models/hdpmodel.py (hunk 8 above)
Comment (tmylk) [inline gensim/models/hdpmodel.py]: Please revert all changes to HdpTopicFormatter

File: gensim/models/hdpmodel.py (hunk 10 above)
Comment (tmylk) [inline gensim/models/hdpmodel.py]: Please revert all changes to HdpTopicFormatter

File: gensim/test/test_hdpmodel.py [hunk 11]
```diff
@@ -47,24 +48,12 @@ def testfile():
     return os.path.join(tempfile.gettempdir(), 'gensim_models.tst')
 
 
-
-class TestHdpModel(unittest.TestCase):
+class TestHdpModel(unittest.TestCase, test_basemodel.TestBaseTopicModel):
     def setUp(self):
         self.corpus = mmcorpus.MmCorpus(datapath('testcorpus.mm'))
         self.class_ = hdpmodel.HdpModel
         self.model = self.class_(corpus, id2word=dictionary)
```
Comment (tmylk) [inline gensim/test/test_hdpmodel.py]: please add the override here

File: gensim/models/hdpmodel.py (hunk 10 above)
Comment (markroxor) [inline gensim/models/hdpmodel.py]: but the Travis build will fail. I have added all the other changes but cannot commit until the tests pass on my local machine.

Comment (tmylk) [inline gensim/models/hdpmodel.py]: what tests fail on your local machine?

File: gensim/test/test_hdpmodel.py (hunk 11 above)
Comment (markroxor) [inline gensim/test/test_hdpmodel.py]: added.

Comment (markroxor) [inline gensim/test/test_hdpmodel.py]: added.

File: gensim/models/hdpmodel.py (hunk 7 above)
Comment (markroxor) [inline gensim/models/hdpmodel.py]: since I've already altered the [HdpTopicFormatter](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/hdpmodel.py#L528) class. To unify the model should I remove it's [print_topic](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/hdpmodel.py#L554) method and inherit it from its parent class `basemodel` instead?

File: gensim/models/basemodel.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+class BaseTopicModel():
```
Comment (piskvorky) [inline gensim/models/basemodel.py]: All Python classes must inherit from object (new-style classes).

File: gensim/models/hdpmodel.py [hunk 13]
```diff
@@ -612,10 +607,9 @@ def show_topic_terms(self, topic_data, num_words):
...
 
+        fmt = (topic_id,fmt)
```
Comment (piskvorky) [inline gensim/models/hdpmodel.py]: PEP8: space after comma.

File: gensim/test/test_basemodel.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+# Copyright (C) 2010 Radim Rehurek <radimrehurek@seznam.cz>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+"""
+Automated tests for checking transformation algorithms (the models package).
+"""
+
+import six
+
+class TestBaseTopicModel():
```
Comment (piskvorky) [inline gensim/test/test_basemodel.py]: All tests should inherit from `unittest.TestCase` (or at least `object`, if you don't want this to be a real test).

Comment (markroxor) [inline gensim/test/test_basemodel.py]: I don't want it to be a real test so adding `object` instead.

File: gensim/models/hdpmodel.py (hunk 13 above)
Comment (markroxor) [inline gensim/models/hdpmodel.py]: added.

File: gensim/models/basemodel.py (hunk 12 above)
Comment (markroxor) [inline gensim/models/basemodel.py]: done.

**Comment by tmylk** [comment]
Hi @markroxor There is a problem with the new tests on windows. It is blocking the release. See https://ci.appveyor.com/project/piskvorky/gensim-431bq/build/job/216bwdmb2gqwkme0

**Comment by markroxor** [comment]
Sent a PR #969
